graceful/service.go

```diff
@@ -76,69 +76,99 @@ func (s *service) ListenAndServeTLS(ctx context.Context, proto string, addr stri
 	return s.listenAndServe(ctx, proto, addr, httpServer)
 }
 
-func (s *service) ListenAndServe(ctx context.Context, proto string, addr string, handler http.Handler) error {
+func (s *Service) ListenAndServe(ctx context.Context, proto string, addr string, handler http.Handler) error {
 	httpServer := &http.Server{
 		Addr:    addr,
 		Handler: handler,
 	}
 	return s.listenAndServe(ctx, proto, addr, httpServer)
 }
 
-func (s *service) listenAndServe(ctx context.Context, proto string, addr string, server *http.Server) error {
+func (s *Service) listenAndServe(ctx context.Context, _ string, addr string, server *http.Server) error {
+	log := logger.Get(ctx)
+
 	if s.pidFile != "" {
 		pid := os.Getpid()
 		err := os.WriteFile(s.pidFile, []byte(fmt.Sprintf("%d\n", pid)), 0600)
 		if err != nil {
-			return errgo.Notef(err, "fail to write PID file")
+			return errors.Wrap(ctx, err, "fail to write PID file")
 		}
 	}
 
-	ld, err := s.graceful.Listen(proto, addr)
+	// Use tableflip to handle graceful restart requests
+	upg, err := tableflip.New(tableflip.Options{
+		UpgradeTimeout: s.reloadWaitDuration,
+		PIDFile:        s.pidFile,
+	})
 	if err != nil {
-		return errgo.Notef(err, "fail to get listener")
+		log.Fatalln(err)
 	}
+	s.graceful = upg
+
+	// setup the signal handling
+	go s.setupSignals(ctx)
+
+	// Listen must be called before Ready
+	ln, err := upg.Listen("tcp", addr)
+	if err != nil {
+		upg.Stop()
+		log.Fatalln("Can't listen:", err)
+	}
+
+	defer upg.Stop()
 
 	if server.TLSConfig != nil {
-		ld = tls.NewListener(ld, server.TLSConfig)
+		ln = tls.NewListener(ln, server.TLSConfig)
 	}
 
 	s.httpServer = server
 
-	go s.setupSignals(ctx)
+	go func() {
+		err := server.Serve(ln)
+		if !errors.Is(err, http.ErrServerClosed) {
+			log.Println("HTTP server:", err)
+		}
+	}()
 
-	err = s.httpServer.Serve(ld)
-	if err == http.ErrServerClosed {
-		return s.waitStopped()
-	}
-	if err != nil {
-		return errgo.Notef(err, "fail to serve http service")
+	log.Printf("ready")
+	if err := upg.Ready(); err != nil {
+		panic(err)
 	}
+	<-upg.Exit()
 
 	// Normally the server should be always gracefully stopped and entering the
 	// above condition when server is closed If by any mean the serve stops
-	// without error, we're stopping the server ourself here.  This code is a
+	// without error, we're stopping the server ourselves here.  This code is a
 	// security to free resource but should be unreachable
 	ctx, cancel := context.WithTimeout(ctx, s.waitDuration)
 	defer cancel()
+	log.Println("shutting down")
 	err = s.shutdown(ctx)
 	if err != nil {
-		return errgo.Notef(err, "fail to shutdown server")
+		return errors.Wrapf(ctx, err, "fail to shutdown server")
 	}
-	return s.waitStopped()
+
+	// Wait for connections to drain.
+	err = server.Shutdown(ctx)
+	if err != nil {
+		log.Println("server shutdown:", err)
```

Comment:
With our logger, we're never using `Println`, we use `Debug` / `Debugf` - `Info` / `Infof` - `Error` / `Errorf` and that's it. 

No `Warn` because it's ambiguous (is it  a problem or not? If yes, it's an error to trigger a Rollbar notification, if not, it's just information, since no one is reviewing `Warn`

`Println` is quite similar to `Info` usually, but less explicit, so we really prefer `Info`